pkg/private/build_tar.py

```diff
@@ -278,34 +277,32 @@ def add_tree(self, tree_top, destpath, mode=None, ids=None, names=None):
             uname=names[0],
             gname=names[1])
 
-  def add_manifest_entry(self, entry_list, file_attributes):
-    entry = manifest.ManifestEntry(*entry_list)
-
+  def add_manifest_entry(self, entry, file_attributes):
     # Use the pkg_tar mode/owner remaping as a fallback
-    non_abs_path = entry.dest.strip('/')
+    non_abs_path = entry["dest"].strip('/')
     if file_attributes:
       attrs = file_attributes(non_abs_path)
     else:
       attrs = {}
     # But any attributes from the manifest have higher precedence
-    if entry.mode is not None and entry.mode != '':
-      attrs['mode'] = int(entry.mode, 8)
-    if entry.user:
-      if entry.group:
-        attrs['names'] = (entry.user, entry.group)
+    if entry["mode"] is not None and entry["mode"] != '':
+      attrs['mode'] = int(entry["mode"], 8)
+    if entry["user"]:
+      if entry["group"]:
+        attrs['names'] = (entry["user"], entry["group"])
       else:
         # Use group that legacy tar process would assign
-        attrs['names'] = (entry.user, attrs.get('names')[1])
-    if entry.entry_type == manifest.ENTRY_IS_LINK:
-      self.add_link(entry.dest, entry.src, **attrs)
-    elif entry.entry_type == manifest.ENTRY_IS_DIR:
-      self.add_empty_dir(entry.dest, **attrs)
-    elif entry.entry_type == manifest.ENTRY_IS_TREE:
-      self.add_tree(entry.src, entry.dest, **attrs)
-    elif entry.entry_type == manifest.ENTRY_IS_EMPTY_FILE:
-      self.add_empty_file(entry.dest, **attrs)
+        attrs['names'] = (entry["user"], attrs.get('names')[1])
+    if entry["type"] == "symlink":
```

Comment:
This may blow out memory usage compared to an int or a reused string.

Reply:
THe more I think about it, the more I think we should not switch to a dict.
named tuples are more efficient and immutable.

Reply:
Be that as it may, how much of a difference will using the named tuple really make?  The performance difference will be nice, but I doubt it will make any noticeable difference compared to the actual I/O cost of building the package. Immutability is also a nice property, but there's nothing stopping scripts from duplicating the data structure and modifying it as needed.

The main benefit of using a dict is that all attributes are in the same bucket -- there's no need to distinguish between what's official and what's an extension.  It's also simpler to maintain.

The Bazel memory use concerns are a better reason.  I tried taking measurements of that recently, and I was not able to get any useful information: results were all over the place and were not sufficiently consistent between runs.  That could be due to the fact that I was working in WSL which may have compatibility issues.

Reply:
I forgot that piece of code just unpacks, consumes and then throws it away. So there really is no memory concern. I guess this is mostly fine.